# SI 330: Data Manipulation 
## 06 - Enhancing pandas peformance

### Dr. Chris Teplovs, School of Information, University of Michigan
<small><a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a>This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.

## Learning Objectives

* explain why we should be concerned with algorithmic efficiency
* be able to describe 5 types of Big O notation
* be able to report execution times in Jupyter
* be able to explain profiling output in Jupyter

**NOTE: You only need to conda install -y line_profiler ONCE!**

In [1]:
!conda install -y line_profiler

Solving environment: ...working... done

## Package Plan ##

  environment location: C:\Users\jodiy\Anaconda3.1

  added / updated specs: 
    - line_profiler


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    line_profiler-2.1.2        |   py36hfa6e2cd_0          73 KB
    conda-4.6.1                |           py36_0         1.7 MB
    ------------------------------------------------------------
                                           Total:         1.8 MB

The following NEW packages will be INSTALLED:

    line_profiler: 2.1.2-py36hfa6e2cd_0

The following packages will be UPDATED:

    conda:         4.5.11-py36_0        --> 4.6.1-py36_0


Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done



line_profiler-2.1.2  | 73 KB     |            |   0% 
line_profiler-2.1.2  | 73 KB     | ########## | 100% 

conda-4.6.1          | 1.7 MB    |            |   0% 
conda-4.6.1          | 1.7 MB    | ##3        |  24% 
conda-4.6.1          | 1.7 MB    | ######4    |  65% 
conda-4.6.1          | 1.7 MB    | ########1  |  82% 
conda-4.6.1          | 1.7 MB    | #########3 |  94% 
conda-4.6.1          | 1.7 MB    | ########## | 100% 


## Big O Notation

* "order of growth"
* a way to categorize how algorithms behave with increasing data size
* commonly used in Computer Science

https://rob-bell.net/2009/06/a-beginners-guide-to-big-o-notation/

### $O(1)$

In [2]:
def IsFirstElementNull(aList):
    return aList[0] == None

### $O(n)$

In [3]:
def ContainsValue(aList, aValue):
    for e in aList:
        if e == aValue:
            return True
    return False

### $O(n^2$)

In [4]:
def ContainsDuplicates(aList):
    for i in range(len(aList)):
        for j in range(len(aList)):
            if i == j:
                continue
            if aList[i] == aList[j]:
                return True
    return False

### $O(2^n$)

In [5]:
def Fibonacci(n):
    if (n <= 1):
        return n
    return Fibonacci(n - 2) + Fibonacci(n -1 )

In [6]:
Fibonacci(10)

55

### $O(log(n))$

In each iteration, we are halfway closer to finishing the problem.

https://stackoverflow.com/questions/2307283/what-does-olog-n-mean-exactly

This is why, for example, looking up people in a phone book is O(log n). You don't need to check every person in the phone book to find the right one; instead, you can simply divide-and-conquer by looking based on where their name is alphabetically, and in every section you only need to explore a subset of each section before you eventually find someone's phone number.

Of course, a bigger phone book will still take you a longer time, but it won't grow as quickly as the proportional increase in the additional size.

We can expand the phone book example to compare other kinds of operations and their running time. We will assume our phone book has businesses (the "Yellow Pages") which have unique names and people (the "White Pages") which may not have unique names. A phone number is assigned to at most one person or business. We will also assume that it takes constant time to flip to a specific page.

Here are the running times of some operations we might perform on the phone book, from best to worst:

O(1) (best case): Given the page that a business's name is on and the business name, find the phone number.

O(1) (average case): Given the page that a person's name is on and their name, find the phone number.

O(log n): Given a person's name, find the phone number by picking a random point about halfway through the part of the book you haven't searched yet, then checking to see whether the person's name is at that point. Then repeat the process about halfway through the part of the book where the person's name lies. (This is a binary search for a person's name.)

O(n): Find all people whose phone numbers contain the digit "5".

O(n): Given a phone number, find the person or business with that number.

O(n log n): There was a mix-up at the printer's office, and our phone book had all its pages inserted in a random order. Fix the ordering so that it's correct by looking at the first name on each page and then putting that page in the appropriate spot in a new, empty phone book.

For the below examples, we're now at the printer's office. Phone books are waiting to be mailed to each resident or business, and there's a sticker on each phone book identifying where it should be mailed to. Every person or business gets one phone book.

O(n log n): We want to personalize the phone book, so we're going to find each person or business's name in their designated copy, then circle their name in the book and write a short thank-you note for their patronage.

O(n2): A mistake occurred at the office, and every entry in each of the phone books has an extra "0" at the end of the phone number. Take some white-out and remove each zero.

O(n · n!): We're ready to load the phonebooks onto the shipping dock. Unfortunately, the robot that was supposed to load the books has gone haywire: it's putting the books onto the truck in a random order! Even worse, it loads all the books onto the truck, then checks to see if they're in the right order, and if not, it unloads them and starts over. (This is the dreaded bogo sort.)

O(nn): You fix the robot so that it's loading things correctly. The next day, one of your co-workers plays a prank on you and wires the loading dock robot to the automated printing systems. Every time the robot goes to load an original book, the factory printer makes a duplicate run of all the phonebooks! Fortunately, the robot's bug-detection systems are sophisticated enough that the robot doesn't try printing even more copies when it encounters a duplicate book for loading, but it still has to load every original and duplicate book that's been printed.

For more mathematical explanation you can checkout how the time complexity arrives to log n here. https://hackernoon.com/what-does-the-time-complexity-o-log-n-actually-mean-45f94bb5bfbf

## Timing and Profiling in Jupyter

The following code and examples are from [Sofia Heisler's repo](https://github.com/s-heisler/pycon2017-optimizing-pandas.git).
It is available at https://github.com/umsi-data-science/s-heisler-pycon2017-optimizing-pandas

In [7]:
import pandas as pd
import numpy as np
from math import *

### Read in the data

In [8]:
df = pd.read_csv('data/new_york_hotels.csv', encoding='cp1252')

In [9]:
df.head()

ean_hotel_id                                     name  \
0        269955       Hilton Garden Inn Albany/SUNY Area   
1        113431     Courtyard by Marriott Albany Thruway   
2        108151                    Radisson Hotel Albany   
3        254756  Hilton Garden Inn Albany Medical Center   
4        198232  CrestHill Suites SUNY University Albany   

                 address1    city state_province postal_code  latitude  \
0     1389 Washington Ave  Albany             NY       12206  42.68751   
1  1455 Washington Avenue  Albany             NY       12206  42.68971   
2             205 Wolf Rd  Albany             NY       12205  42.72410   
3     62 New Scotland Ave  Albany             NY       12208  42.65157   
4  1415 Washington Avenue  Albany             NY       12206  42.68873   

   longitude  star_rating  high_rate  low_rate  
0  -73.81643          3.0   154.0272  124.0216  
1  -73.82021          3.0   179.0100  134.0000  
2  -73.79822          3.0   134.1700   84.1600  
3  -73.77638          3.0   308.2807  228.4597  
4  -73.81854          3.0   169.3900   89.3900

In [10]:
df.columns

Index(['ean_hotel_id', 'name', 'address1', 'city', 'state_province',
       'postal_code', 'latitude', 'longitude', 'star_rating', 'high_rate',
       'low_rate'],
      dtype='object')

## Benchmarking example

#### Define the normalization function

In [2]:
def normalize(df, pd_series):
    pd_series = pd_series.astype(float)
    #print(pd_series)
    # Find upper and lower bound for outliers
    avg = np.mean(pd_series)
    sd  = np.std(pd_series)
    #print(avg,sd)
    lower_bound = avg - 2*sd
    upper_bound = avg + 2*sd

    # Collapse in the outliers
    df.loc[pd_series < lower_bound , "high_rate" ] = lower_bound
    df.loc[pd_series > upper_bound , "high_rate" ] = upper_bound
    # Finally, take the log (why add 1?)
    normalized_price = np.log(df["high_rate"].astype(float)+1)
    return normalized_price

In [3]:
normalize(df, df["high_rate"])

NameError: name 'df' is not defined

#### Timing the normalization function

In [26]:
%timeit df["high_rate_normalized"] = normalize(df, df["high_rate"])

9.41 ms ± 1.16 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [20]:
%timeit x = 10

28.4 ns ± 0.163 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


#### Profiling the normalization function

In [27]:
    %load_ext line_profiler

The line_profiler extension is already loaded. To reload it, use:
  %reload_ext line_profiler


In [1]:
%lprun -f normalize df["high_rate_normalized"] = normalize(df, df["high_rate"])

UsageError: Line magic function `%lprun` not found.


### <font color="magenta"> Q1 (2 points): Where is most of the time being spent?

The normalize function when calculating the upper and lower bounds. Lines 12 and 13

### <font color="magenta"> Q2 (2 points): What order is the normalize function?  How can you figure this out?

O(1). Order one uses pre-determined indices and this is the fastest way.

## There are at least 5 different ways to operate on a pandas Series:
* iterate with a for loop
* use iterrows
* use apply
* use vectorized functions
* use cython

## Haversine definition

https://en.wikipedia.org/wiki/Haversine_formula

In [29]:
def haversine(lat1, lon1, lat2, lon2):
    miles_constant = 3959
    lat1, lon1, lat2, lon2 = map(np.deg2rad, [lat1, lon1, lat2, lon2])
    dlat = lat2 - lat1 
    dlon = lon2 - lon1 
    a = np.sin(dlat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2)**2
    c = 2 * np.arcsin(np.sqrt(a)) 
    mi = miles_constant * c
    return mi

## Iterating using for and loc

### <font color="magenta"> Q3 (2 points): Write pythonic code that iterates over each row to calculate the distance from lat1= 40.671, lon1= -73.985 to each hotel's latitude and longitude.

In [33]:
%%timeit
for i in range(len(df)):
    row = df.iloc[i]
    h = (haversine(40.671, -73.985,row['latitude'], row['longitude']))

564 ms ± 73.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


## Iterrows Haversine

In [34]:
%%timeit
# Haversine applied on rows via iteration
haversine_series = []
for index, row in df.iterrows():
    haversine_series.append(haversine(40.671, -73.985,\
                                      row['latitude'], row['longitude']))
df['distance'] = haversine_series

295 ms ± 7.45 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


## Apply Haversine on rows

### Timing "apply"

In [35]:
%timeit df['distance'] =\
df.apply(lambda row: haversine(40.671, -73.985,\
                               row['latitude'], row['longitude']), axis=1)

126 ms ± 6.13 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


#### Profiling "apply"

In [37]:
# Haversine applied on rows
%lprun -f haversine \
df.apply(lambda row: haversine(40.671, -73.985,\
                               row["latitude"], row["longitude"]), axis=1)

Timer unit: 3.95061e-07 s

Total time: 0.0561271 s
File: <ipython-input-29-ad000f5aaf6c>
Function: haversine at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
     1                                           def haversine(lat1, lon1, lat2, lon2):
     2      1631       3879.0      2.4      2.7      miles_constant = 3959
     3      1631      39749.0     24.4     28.0      lat1, lon1, lat2, lon2 = map(np.deg2rad, [lat1, lon1, lat2, lon2])
     4      1631       4006.0      2.5      2.8      dlat = lat2 - lat1 
     5      1631       2961.0      1.8      2.1      dlon = lon2 - lon1 
     6      1631      57959.0     35.5     40.8      a = np.sin(dlat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2)**2
     7      1631      26250.0     16.1     18.5      c = 2 * np.arcsin(np.sqrt(a)) 
     8      1631       4747.0      2.9      3.3      mi = miles_constant * c
     9      1631       2521.0      1.5      1.8      return mi

## Vectorized implementation of Haversine applied on Pandas series

#### Timing vectorized implementation

In [38]:
# Vectorized implementation of Haversine applied on Pandas series
%timeit df['distance'] = haversine(40.671, -73.985,\
                                   df['latitude'], df['longitude'])

3.28 ms ± 245 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


#### Profiling vectorized implementation

In [40]:
# Vectorized implementation profile
%lprun -f haversine haversine(40.671, -73.985,\
                              df["latitude"], df["longitude"])

Timer unit: 3.95061e-07 s

Total time: 0.00597135 s
File: <ipython-input-29-ad000f5aaf6c>
Function: haversine at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
     1                                           def haversine(lat1, lon1, lat2, lon2):
     2         1          7.0      7.0      0.0      miles_constant = 3959
     3         1        791.0    791.0      5.2      lat1, lon1, lat2, lon2 = map(np.deg2rad, [lat1, lon1, lat2, lon2])
     4         1        775.0    775.0      5.1      dlat = lat2 - lat1 
     5         1        848.0    848.0      5.6      dlon = lon2 - lon1 
     6         1       8348.0   8348.0     55.2      a = np.sin(dlat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2)**2
     7         1       3087.0   3087.0     20.4      c = 2 * np.arcsin(np.sqrt(a)) 
     8         1       1254.0   1254.0      8.3      mi = miles_constant * c
     9         1          5.0      5.0      0.0      return mi

## Vectorized implementation of Haversine applied on NumPy arrays

#### Timing vectorized implementation

In [41]:
# Vectorized implementation of Haversine applied on NumPy arrays
%timeit df['distance'] = haversine(40.671, -73.985,\
                         df['latitude'].values, df['longitude'].values)

523 µs ± 4.13 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [42]:
%%timeit
# Convert pandas arrays to NumPy ndarrays
np_lat = df['latitude'].values
np_lon = df['longitude'].values

10.6 µs ± 391 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


#### Profiling vectorized implementation

In [44]:
%lprun -f haversine df["distance"] = haversine(40.671, -73.985,\
                        df["latitude"].values, df["longitude"].values)

Timer unit: 3.95061e-07 s

Total time: 0.00110103 s
File: <ipython-input-29-ad000f5aaf6c>
Function: haversine at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
     1                                           def haversine(lat1, lon1, lat2, lon2):
     2         1          8.0      8.0      0.3      miles_constant = 3959
     3         1        431.0    431.0     15.5      lat1, lon1, lat2, lon2 = map(np.deg2rad, [lat1, lon1, lat2, lon2])
     4         1        632.0    632.0     22.7      dlat = lat2 - lat1 
     5         1         47.0     47.0      1.7      dlon = lon2 - lon1 
     6         1       1478.0   1478.0     53.0      a = np.sin(dlat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2)**2
     7         1        168.0    168.0      6.0      c = 2 * np.arcsin(np.sqrt(a)) 
     8         1         20.0     20.0      0.7      mi = miles_constant * c
     9         1          3.0      3.0      0.1      return mi

### <font color="magenta"> Q4 (2 points): Record the times above and comment on how efficiency improves going from the first implementation to the last

564 ms using for & iloc
295 ms using iterrows
126 ms using apply
526 microseconds using vectorized functions
10.6 microseconds using vectorized functions on numpy arrays

Overall improvement of factor of 1000 or so.523 µs ± 4.13 µs per loop

## Not good enough?  Cythonize that loop

#### Load the cython extension

In [1]:
%load_ext cython

#### Run unaltered Haversine through Cython

In [2]:
%%cython -a

# Haversine cythonized (no other edits)
import numpy as np
cpdef haversine_cy(lat1, lon1, lat2, lon2):
    miles_constant = 3959
    lat1, lon1, lat2, lon2 = map(np.deg2rad, [lat1, lon1, lat2, lon2])
    dlat = lat2 - lat1 
    dlon = lon2 - lon1 
    a = np.sin(dlat/2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon/2)**2
    c = 2 * np.arcsin(np.sqrt(a))    
    mi = miles_constant * c
    return mi

#### Time it

In [47]:
%timeit df['distance'] =\
       df.apply(lambda row: haversine_cy(40.671, -73.985,\
                row['latitude'], row['longitude']), axis=1)

128 ms ± 8.39 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


### <font color="magenta"> Q5 (2 points): How much more efficient is the cython version compared to the best non-cython version from above?

Cython version was more efficient but not by a lot.

#### Redefine Haversine with data types and C libraries

### FYI Only

In [ ]:
%%cython -a
# Haversine cythonized
from libc.math cimport sin, cos, acos, asin, sqrt

cdef deg2rad_cy(float deg):
    cdef float rad
    rad = 0.01745329252*deg
    return rad
    
cpdef haversine_cy_dtyped(float lat1, float lon1, float lat2, float lon2):
    cdef: 
        float dlon
        float dlat
        float a
        float c
        float mi
    
    lat1, lon1, lat2, lon2 = map(deg2rad_cy, [lat1, lon1, lat2, lon2])
    dlat = lat2 - lat1 
    dlon = lon2 - lon1 
    a = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
    c = 2 * asin(sqrt(a)) 
    mi = 3959 * c
    return mi


#### Time it

In [ ]:
%timeit df['distance'] =\
df.apply(lambda row: haversine_cy_dtyped(40.671, -73.985,\
                              row['latitude'], row['longitude']), axis=1)

# END OF NOTEBOOK
Please remember to submit your notebook in .ipynb and .html formats.